src/blog/oilpan-pointer-compression.md

```diff
@@ -0,0 +1,247 @@
+
+## Background
+
+Pointer compression is a mechanism to reduce the size of pointers on 64-bit platforms.  Pointers in Oilpan are encapsulated in a smart pointer called [`Member`](https://source.chromium.org/chromium/chromium/src/+/main:v8/include/cppgc/member.h). In an uncompressed heap layout, `Member` references directly point to heap objects, i.e., 8 bytes of memory are used per reference. In such a scenario, the heap may be spread out over the whole address space as each pointer contains all the relevant information to refer to an object.
+
+![Uncompressed heap layout](/_img/oilpan-pointer-compression/uncompressed-layout.svg)
+
+With a compressed heap layout, `Member` references are only offsets into a heap cage, which is a contiguous region of memory.  The combination of a base pointer (base) that points to the beginning of the heap cage and a Member forms a full pointer, very similar to how [segmented addressing](https://en.wikipedia.org/wiki/Memory_segmentation#Segmentation_without_paging) works.  The size of a heap cage is limited by the available bits for the offset. E.g., a 4GB heap cage requires 32-bit offsets.
+
+![Compressed heap layout](/_img/oilpan-pointer-compression/compressed-layout.svg)
+
+Conveniently, Oilpan heaps are already contained in such a 4GB heap cage on 64-bit platforms, to allow referring to garbage collection metadata by just aligning any valid heap pointer down to the closest 4GB boundary.
+
+Oilpan also supports multiple heaps in the same process to, e.g., support web workers with their own C++ heaps in Blink.  The problem arising from this setup is how to map heaps to possibly many heap cages.  Since heaps are bound to native threads in Blink, the solution at hand here is to refer to heap cages via a thread-local base pointer.  Depending on how V8 and its embedders are compiled, the thread local storage (TLS) model can be restricted to speed up how the base is loaded from memory.  Ultimately, the most generic TLS mode is required though to support Android, as on this platform the renderer (and thus V8) are loaded via `dlopen`.  It is such restrictions that make the use of TLS infeasible from a performance perspective[^1].  In order to provide best performance, Oilpan, similar to V8, allocates all heaps into a single heap cage when using pointer compression.  While this does restrict overall memory available, we believe that this is currently acceptable given that pointer compression already aims at reducing memory.  If a single 4GB heap cage proves to be too restrictive, the current compression scheme allows to increase the size of the heap cage size to 16GB without sacrificing performance.
```

Comment:
The last sentence, please change:
"... to increase the size of the heap cage size..."
to
"... to increase the heap cage size..."

Reply:
Tip: GitHub supports [Suggestions](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#:~:text=Optionally%2C%20to%20suggest%20a%20specific%20change%20to%20the%20line%20or%20lines%2C%20click%20%2C%20then%20edit%20the%20text%20within%20the%20suggestion%20block.) Example:

```suggestion
Oilpan also supports multiple heaps in the same process to, e.g., support web workers with their own C++ heaps in Blink.  The problem arising from this setup is how to map heaps to possibly many heap cages.  Since heaps are bound to native threads in Blink, the solution at hand here is to refer to heap cages via a thread-local base pointer.  Depending on how V8 and its embedders are compiled, the thread local storage (TLS) model can be restricted to speed up how the base is loaded from memory.  Ultimately, the most generic TLS mode is required though to support Android, as on this platform the renderer (and thus V8) are loaded via `dlopen`.  It is such restrictions that make the use of TLS infeasible from a performance perspective[^1].  In order to provide best performance, Oilpan, similar to V8, allocates all heaps into a single heap cage when using pointer compression.  While this does restrict overall memory available, we believe that this is currently acceptable given that pointer compression already aims at reducing memory.  If a single 4GB heap cage proves to be too restrictive, the current compression scheme allows to increase the heap cage size to 16GB without sacrificing performance.
```